docs/user/developing_jobs.md

```diff
@@ -42,21 +44,115 @@ In brief, the following general steps can be followed:
+
+Here is a better approach that utilizes diffsync's `get_or_instantiate` and Django's [`select_related`](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#select-related) for query optimization purposes:
+
+```python
+def load(self):
+    # This next lines represents the single (!) database query that replaces the 7033 from the previous example
+    for location in Location.objects.all().select_related("site", "site__region", "site__region__parent"):
+        parent_region, parent_region_created = self.get_or_instantiate("parent_region", ids={"name": location.site.region.parent.name})
+        if parent_region_created:
+            self.add(parent_region)
+        child_region, child_region_created = self.get_or_instantiate("child_region", ids={"name": location.site.region.parent})
+        if child_region_created:
+            self.add(child_region)
+            parent_region.add_child(child_region)
+        site, site_created = self.get_or_instantiate("site", ids={"name": location.site.name})
+        if site_created:
+            self.add(site)
+            child_region.add_child(site)
+        location, location_created = self.get_or_instantiate("location", ids={"name": location.name})
+        if location_created:
+            self.add(location)
+            site.add_child(location)
+```
+
+As an additional bonus, this way the code has less levels of indentation.
+
+The essence of this is that you should make liberal use of `select_related` to join together the database tables you need into a single, big query rather than a bunch of small queries.
+
+!!! note
+    Check out the [Django documentation](https://docs.djangoproject.com/en/3.2/topics/db/optimization/) for a more comprehensive source on optimizing database access.
+
+### Optimizing worker stdout IO
+
+If after optimizing your database access you are still facing performance issues, you should check out the [analyzing job performance](#analyzing-job-performance) section of the docs. Should you find that a certain `io.write` appears high up in the ranking, you are probably facing an issue where your job is writing to stdout so quickly that your worker node/process cannot drain its buffer quickly enough. To deal with this, tone down on what you are logging to stdout inside your job. This could be any of the following things (non-exhaustive, check out your worker logs):
+
+- diffsync [logging configuration](https://diffsync.readthedocs.io/en/latest/api/diffsync.logging.html?highlight=logging)
+- `print` calls
+- other, external frameworks you are using
+
+## Analyzing Job Performance
+
+In general there are two different metrics to optimize for when developing SSoT jobs:
+
+- CPU time (maps directly to total execution time)
+- Memory usage
```

Comment:
Should IO be on this list?

Reply:
I don't think IO is something you optimize for, stdout IO is something you can do less of to optimize for CPU time, isn't it?

Reply:
One of the issues I was told was basically the same premise you described in the example but for external systems. e.g. Make an API call to each device, interface, asset information in ServiceNow, vs making a paginated call to Devices, Interfaces, and Assets, (so only 3 paginated calls vs N).

Reply:
Ohh I understand. Yes that is 100% worth calling out, too. Thanks!

Reply:
Also, it would be good to provide some ideas to improve, with the tradeoffs (i.e. overloading the target endpoint)

Reply:
You could use threading or asyncio to parallelize IO, but this could overload the other endpoint.

Reply:
https://github.com/nautobot/nautobot-plugin-ssot/pull/104#issuecomment-1531198525

Reply:
Addressed